steel/src/contract.rs

```diff
@@ -0,0 +1,441 @@
+    pub fn from(mut self, from: Address) -> Self {
+        self.caller = from;
+        self
+    }
+
+    /// Sets the gas limit of the function call.
+    pub fn gas(mut self, gas: u64) -> Self {
+        self.gas_limit = gas;
+        self
+    }
+
+    /// Sets the gas price of the function call.
+    pub fn gas_price(mut self, gas_price: U256) -> Self {
+        self.gas_price = gas_price;
+        self
+    }
+
+    /// Sets the value field of the function call.
+    pub fn value(mut self, value: U256) -> Self {
+        self.value = value;
+        self
+    }
+
+    /// Executes the view call using the given environment.
+    #[inline]
+    #[deprecated(since = "0.11.0", note = "please use `Contract::new` instead")]
+    pub fn execute<H: EvmHeader>(self, env: GuestViewCallEnv<H>) -> C::Return {
+        ViewCallBuilder {
+            transaction: self,
+            env: &env,
+        }
+        .call()
+    }
+
+    /// Executes the call for the provided state.
+    pub(crate) fn transact<'a, E>(self, env: E) -> Result<C::Return, String>
+    where
+        C: SolCall,
+        E: private::SteelEnv<'a> + 'a,
+        <<E as private::SteelEnv<'a>>::Db as revm::Database>::Error: Debug,
+    {
+        let mut evm = env.into_evm();
+
+        let tx_env = evm.tx_mut();
+        tx_env.caller = self.caller;
+        tx_env.gas_limit = self.gas_limit;
+        tx_env.gas_price = self.gas_price;
+        tx_env.transact_to = TransactTo::call(self.contract);
+        tx_env.value = self.value;
+        tx_env.data = self.data.into();
+
+        let ResultAndState { result, .. } = evm
+            .transact_preverified()
+            .map_err(|err| format!("Call '{}' failed: {:?}", C::SIGNATURE, err))?;
+        let ExecutionResult::Success { reason, output, .. } = result else {
+            return Err(format!("Call '{}' failed", C::SIGNATURE));
+        };
+        // there must be a return value to decode
+        if reason != SuccessReason::Return {
+            return Err(format!(
+                "Call '{}' did not return: {:?}",
+                C::SIGNATURE,
+                reason
+            ));
+        }
+        let returns = C::abi_decode_returns(&output.into_data(), true).map_err(|err| {
+            format!(
+                "Call '{}' returned invalid type; expected '{}': {:?}",
+                C::SIGNATURE,
+                <C::ReturnTuple<'_> as SolType>::SOL_NAME,
+                err
+            )
+        })?;
+
+        Ok(returns)
+    }
+}
+
+mod private {
+    use super::*;
+    use crate::{MerkleTrie, StateDB};
+
+    pub trait SteelEnv<'a> {
+        type Db: revm::Database;
+
+        fn into_evm(self) -> Evm<'a, (), Self::Db>
+        where
+            <Self::Db as Database>::Error: Debug;
+    }
+
+    impl<'a, H> SteelEnv<'a> for &'a GuestViewCallEnv<H>
+    where
+        H: EvmHeader,
+    {
+        type Db = WrapStateDb<'a>;
+
+        fn into_evm(self) -> Evm<'a, (), Self::Db> {
```

Comment:
In the past, the EVM builder was very likely to change in newer version also its usage is error prone. So I would rather deduplicate that part, especially since the only change is the wrapper. So personally I'd get rid of `SteelEnv`, we have two implementations of `call` anyway, so maybe we can make that distinction there.

Reply:
yeah fair, I was hoping these would be more easily generic, but that does seem like a better path. I'll try that again, and either remember why I didn't go that way or I'll make that change.

Reply:
Good callout! So originally it was needed because the call function was the same, but after that change was made, I forgot to loop back to remove :) done in [`c7f52fc` (#115)](https://github.com/risc0/risc0-ethereum/pull/115/commits/c7f52fc9374fc6414d17a10a0bd31ed2c21a0779)